Source/Pawnmorphs/Esoteria/Hediffs/Base/Hediff_PM.cs

```diff
@@ -0,0 +1,543 @@
+        /// <see cref="LoadSaveMode.PostLoadInit" /> phase).
+        /// <br />
+        /// Any initialization that needs to be repeated after every load can be done here.
+        /// </summary>
+        protected virtual void PostAddOrLoad()
+        {
+            HashOffset = pawn?.HashOffset() ?? 0;
+        }
+
+        /// <summary>
+        /// Called after the hediff is removed from the pawn.
+        /// </summary>
+        public override void PostRemoved()
+        {
+            base.PostRemoved();
+            foreach (HediffComp_PM pmComp in PMComps)
+                pmComp.CompPostPostRemoved();
+        }
+
+        //TODO - override Tick() the same way Hediff_StageChanges does
+
+        /// <summary>
+        /// Called after the main Hediff tick.  Comps and PMComps are ticked in this method.
+        /// <br />
+        /// Note that this method is a MAJOR hot path -- it should be optimized as much as possible because it potentially gets
+        /// called tens of thousands of times per second. For this reason, foreach loops should be avoided because the iterator
+        /// object creation is a notable source of slowdown.
+        /// </summary>
+        public override void PostTick()
+        {
+            // We're not calling base.PostTick() here because we're manually replicating it in a slightly more optimized way
+            var severityAdjustment = 0f;
+
+            if (_hasComps && comps != null)
+            {
+                int compCount = comps.Count; // Not caching this genuinely has a noticeable performance impact for hediffs that
+                // have multiple comps.  We can't cache it beyond the tick, however, because
+                // exceptions in certain places can cause comps to be removed.
+                // ReSharper disable once ForCanBeConvertedToForeach
+                for (var index = 0; index < compCount; ++index)
+                    comps[index]?.CompPostTick(ref severityAdjustment);
+            }
+
+            if (_hasPMComps)
+            {
+                int compCount = _pmComps.Count; // Same as above
+                int hashOffsetTick = Find.TickManager!.TicksGame + HashOffset;
+
+                // Note that PMComps.CompPostTick() is never called.  This is because that method exists only for legacy support.
+                // That method cannot be overridden and only does exactly what we're doing here (but slightly less efficiently)
+
+                if (hashOffsetTick % 60 == 0) // Every real-life second
+                    for (var index = 0; index < compCount; ++index)
+                        _pmComps[index]?.TickSecond(ref severityAdjustment);
+
+                if (hashOffsetTick % 2500 == 0) // Once an in-game hour
+                    for (var index = 0; index < compCount; ++index)
+                        _pmComps[index]?.TickHour(ref severityAdjustment);
+
+                if (hashOffsetTick % 60000 == 0) // Once an in-game day
```

Comment:
This could technically be gated behind % 60 checks to only have 2 modulus checks per tick.

Reply:
It could, yeah.  I figured it was more readable this way, and I haven't tested it to see if the performance matters enough yet.